geomstats/learning/riemannian_mean_shift.py

```diff
@@ -0,0 +1,161 @@
+        self.n_centers = n_centers
+        self.n_jobs = n_jobs
+        self.max_iter = max_iter
+        self.init_centers = init_centers
+        self.kernel = kernel
+        self.centers = None
+
+        print(type(max_iter))
+        print(max_iter)
+
+    def __intersets_distances(self, points_A, points_B, **joblib_kwargs):
+        """Parallel computation of distances between two sets of points"""
+        n_A, n_B = points_A.shape[0], points_B.shape[0]
+
+        @joblib.delayed
+        @joblib.wrap_non_picklable_objects
+        def pickable_dist(x, y):
+            """Riemannian metric between points x & y"""
+            return self.metric.dist(x, y)
+
+        pool = joblib.Parallel(n_jobs=self.n_jobs, **joblib_kwargs)
+        out = pool(
+            pickable_dist(points_A[i, :], points_B[j, :])
+            for i in range(n_A)
+            for j in range(n_B)
+        )
+
+        return gs.array(out).reshape((n_A, n_B))
+
+    def fit(self, X):
+        """Fit centers in all the input points to find 'n_centers' number of clusters"""
+
+        @joblib.delayed
+        @joblib.wrap_non_picklable_objects
+        def pickable_mean(points, weights):
+            """Frechet Mean of all points weighted by weights"""
+            return self.mean.fit(points, weights=weights).estimate_
+
+        # 'from_points' is the default initialization for centers used
+        if self.init_centers == "from_points":
+            n_points = X.shape[0]
+            centers = X[gs.random.randint(n_points, size=self.n_centers), :]
+        if self.init_centers == "random_uniform":
+            n_centers = centers.shape[0]
+            centers = self.manifold.random_uniform(n_samples=n_centers)
+
+        print(type(self.max_iter))
+        print(self.max_iter)
+        for _ in range(self.max_iter):
+            dists = self.__intersets_distances(centers, X)
+
+            # assuming the use of 'flat' kernel
+            if self.kernel == "flat":
+                weights = gs.ones_like(dists)
+
+            weights[dists > self.bandwidth] = 0.0
+            weights = weights / gs.sum(weights, axis=1, keepdims=1)
+
+            points_to_average, nonzero_weights = [], []
+
+            for j in range(self.n_centers):
+                indexes = gs.where(weights[j] > 0)
+                nonzero_weights += [
+                    weights[j][indexes],
+                ]
+                points_to_average += [
+                    X[indexes],
+                ]
+
+            # compute Frechet means in parallel
+            pool = joblib.Parallel(n_jobs=self.n_jobs)
+            out = pool(
+                pickable_mean(points_to_average[j], nonzero_weights[j])
+                for j in range(self.n_centers)
+            )
+
+            new_centers = gs.array(out)
+
+            displacements = [self.metric.dist(centers[j], new_centers[j])]
+            centers = new_centers
+
+            if (gs.array(displacements) < self.tol).all():
```

Comment:
Yes, you might not need the list (see comment above) especially if you are re-casting to array then. Try removing the list `[]` on line 141 and the `gs.array` here

Reply:
@shubhamtalbar96 removing the list didn't work?